DEVELOPER.md

```diff
@@ -0,0 +1,305 @@
+````
+target_link_libraries(kokkoscore PUBLIC Kokkos::HWLOC)
+````
+The complexity of which includes, options, and libraries the TPL requires
+should be encapsulated in the CMake target.
+
+#### TPL Helper Functions
+##### KOKKOS_IMPORT_TPL
+This function can be invoked as, e.g.
+````
+KOKKOS_IMPORT_TPL(HWLOC)
+````
+This function checks if the TPL was enabled by a `-DKokkos_ENABLE_HWLOC=On` flag.
+If so, it calls `find_package(TPLHWLOC)`.
+This invokes the file `FindTPLHWLOC.cmake` which should be contained in the `cmake/Modules` folder.
+If successful, another function `KOKKOS_EXPORT_CMAKE_TPL` gets invoked.
+This automatically adds all the necessary import commands to `KokkosConfig.cmake`.
+
+##### KOKKOS_FIND_IMPORTED
+Inside a `FindTPLX.cmake` file, the simplest way to import a library is to call, e.g.
+````
+KOKKOS_FIND_IMPORTED(HWLOC LIBRARY hwloc HEADER hwloc.h)
+````
+This finds the location of the library and header and creates an imported target `Kokkos::HWLOC`
+that can be linked against.
+The library/header find can be guided with `-DHWLOC_ROOT=` or `-DKokkos_HWLOC_DIR=` during CMake configure.
+These both specify the install prefix.
+
+##### KOKKOS_LINK_TPL
+This function checks if the TPL has been enabled.
+If so, it links a given library against the imported (or interface) TPL target.
+
+##### KOKKOS_CREATE_IMPORTED_TPL
+This helper function is best understood by reading the actual code.
+This function takes arguments specifying the properties and creates the actual TPL target.
+The most important thing to understand for this function is whether you call this function with the optional `INTERFACE` keyword.
+This tells the project to either create the target as an imported target or interface target, as discussed above.
+
+##### KOKKOS_EXPORT_CMAKE_TPL
+Even if the TPL just loads a valid CMake target, we still must "export" it into the config file.
+When Kokkos is loaded by a downstream project, this TPL must be loaded.
+Calling this function simply appends text recording the location where the TPL was found
+and adding a `find_dependency(...)` call that will reload the CMake target.
+
+### The Great TriBITS Compromise
+
+TriBITS was a masterpiece of CMake version 2 before the modern CMake idioms of building and using.
+TriBITS greatly limited verbosity of CMake files, handled complicated dependency trees between packages, and handled automatically setting up include and linker paths for dependent libraries.
+
+Kokkos is now used by numerous projects that don't (and won't) depend on TriBITS for their build systems.
+Kokkos has to work outside of TriBITS and provide a standard CMake 3+ build system.
+At the same time, Kokkos is used by numerous projects that depend on TriBITS and don't (and won't) switch to a standard CMake 3+ build system.
+
+Nevertheless, Kokkos must satisfy all customers for now.
+Kokkos is implemented as something TriBITS-like.
+Instead of calling functions `TRIBITS_X(...)` it calls a function `KOKKOS_X(...)`.
```

Comment:
```suggestion
This is achieved by providing an abstraction layer for `TriBITS` functions:
```

Reply:
Reworded - thought not exactly as suggested.